### PR TITLE
Prune SAM3 tracker memory to prevent VRAM leak

### LINKED ISSUE #22950: 显存泄露
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

_No response_

### Bug

使用yolo11最新集成的sam3模型进行视频文本追踪时，显存越用越多是怎么回事，在服务器上从5G一直往上涨，最后48G显存都不够用，即使打开了stream=True，本地也是同样的情况

### Environment

(yolo11) PS E:\A_worker\W_XM\yolo11_car_seg\yolo_sam3> yolo checks
Ultralytics 8.3.237  Python-3.12.12 torch-2.9.1+cu128 CUDA:0 (NVIDIA GeForce RTX 3050 Ti Laptop GPU, 8192MiB)
Setup complete  (20 CPUs, 31.7 GB RAM, 754.2/953.9 GB disk)

OS                     Windows-11-10.0.26100-SP0
Environment            Windows
Python                 3.12.12
Install                pip
Path                   E:\Miniconda\envs\yolo11\Lib\site-packages\ultralytics
RAM                    31.71 GB
Disk                   754.2/953.9 GB
CPU                    12th Gen Intel Core(TM) i7-12700H
CPU count              20
GPU                    NVIDIA GeForce RTX 3050 Ti Laptop GPU, 8192MiB
GPU count              1
CUDA                   12.8

numpy                   2.2.6>=1.23.0
matplotlib              3.10.7>=3.3.0
opencv-python           4.12.0.88>=4.6.0
pillow                  11.3.0>=7.1.2
pyyaml                  6.0.3>=5.3.1
requests                2.32.5>=2.23.0
scipy                   1.16.3>=1.4.1
torch                   2.9.1+cu128>=1.8.0
torch                   2.9.1+cu128!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision             0.24.1+cu128>=0.9.0
psutil                  7.1.3>=5.8.0
polars                  1.36.0>=0.20.0
ultralytics-thop        2.0.18>=2.0.18
(yolo11) PS E:\A_worker\W_XM\yolo11_car_seg\yolo_sam3> 

### Minimal Reproducible Example

import cv2
from ultralytics.models.sam import SAM3VideoSemanticPredictor

# Initialize semantic video predictor
overrides = dict(conf=0.25,
                 task="segment",
                 mode="predict",
                 imgsz=640,
                 model=r"E:\A_worker\W_XM\yolo11_sam3\models\sam3\sam3.pt",
                 half=True,
                 vid_stride=10,)
predictor = SAM3VideoSemanticPredictor(overrides=overrides,
                                       bpe_path=r"E:\A_worker\W_XM\yolo11_sam3\assets\bpe_simple_vocab_16e6.txt.gz")

# Track concepts using text prompts
results = predictor(source=r"E:\A_worker\W_XM\yolo11_car_seg\test_img\flow\2.mp4",
                    text=["vehicle"],
                    stream=True,
                    save=True)

# Process results
for r in results:
    img = r.plot()  # 处理当前帧，生成带追踪框的图像
    cv2.imshow('yolo_sam3', img)

    # 关键修改：将参数从 0 改为 1。程序会等待1毫秒并自动继续。
    # 同时检测是否按下了 'q' 键，以便随时退出。
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cv2.destroyAllWindows()

# Alternative: Track with bounding box prompts
# results = predictor(
#     source="path/to/video.mp4",
#     bboxes=[[864, 383, 975, 620], [705, 229, 782, 402]],
#     labels=[1, 1],  # Positive labels
#     stream=True,
#     save=True,
# )

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes #22950 

**MRE**

```python
from ultralytics.models.sam import SAM3VideoSemanticPredictor

# Initialize semantic video predictor
overrides = dict(
    conf=0.25, 
    task="segment", 
    mode="predict", 
    imgsz=640, 
    model="sam3.pt", 
    half=True, 
    device=0
)
predictor = SAM3VideoSemanticPredictor(overrides=overrides)

# Track concepts using text prompts
results = predictor(source="video1.mp4", text=["person"], stream=True, save=False)

# Process results
import torch

for i, r in enumerate(results):
    if torch.cuda.is_available():
        print(i, "MB alloc=", torch.cuda.memory_allocated() / 1024**2,
                 "MB max=", torch.cuda.max_memory_allocated() / 1024**2)
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds automatic pruning of SAM2 “non-conditioning” frame memory to keep video inference memory usage bounded 🧠🧹

### 📊 Key Changes
- Enables a new default flag `self.clear_non_cond_mem = True` to periodically clear old non-conditioning memory ✅
- Calls a new helper `_prune_non_cond_memory(...)` after each frame’s output is stored during:
  - Standard `inference(...)`
  - Video `propagate_in_video(...)` 🎥
- Implements `_prune_non_cond_memory` to delete outdated entries from:
  - Global `output_dict["non_cond_frame_outputs"]`
  - Per-object `output_dict_per_obj[*]["non_cond_frame_outputs"]` 🗂️
- Uses the model’s temporal memory parameters (`num_maskmem` and `memory_temporal_stride_for_eval`) to decide how far back to retain frames ⏱️

### 🎯 Purpose & Impact
- Reduces RAM/VRAM growth over long videos by preventing unbounded accumulation of cached frame outputs 📉
- Improves stability for extended tracking/segmentation runs (fewer slowdowns and memory-related crashes) 🛡️
- Keeps recent context needed for quality while discarding older, less useful non-conditioning history ⚖️
- Minimal user-facing changes: behavior improves by default without requiring new configuration 🙌

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). This is an automated message 🤖—an Ultralytics engineer will assist shortly. Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Overall the PR moves in the right direction by bounding per-frame/per-object history and cleaning up per-object state on removal. Main issues to address are robustness of frame-key comparisons in `_prune_memory_bank()` (avoid `TypeError` if keys aren’t ints) and a mismatch between `_consolidate_tracker_states()`’s intent/docstring and its current behavior (unused `max_states` and no actual capping). Also consider unifying the score-prune window with `memory_bank_max_frames` to avoid divergent retention policies.

💬 Posted 4 inline comments

File: ultralytics/models/sam/predict.py [hunk 1]
```diff
@@ -3920,3 +3928,39 @@ def _drop_new_det_with_obj_limit(self, new_det_fa_inds, det_scores_np, num_to_ke
...
+
+        min_frame = frame_idx - self.memory_bank_max_frames
+
+        for state in tracker_states_local:
+            output_dict = state.get("output_dict", {})
+
+            # only prune non_cond_frame_outputs
+            storage_key = "non_cond_frame_outputs"
+            if storage_key in output_dict:
+                frames_to_remove = [k for k in output_dict[storage_key] if k < min_frame]
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/predict.py]: ⚠️ **HIGH**: `frames_to_remove` assumes all keys are ints. If the dict ever contains non-int keys (e.g., strings/sentinels), `k < min_frame` will raise `TypeError`. Safer to filter by `isinstance(k, int)` like you did in `_prune_frame_scores()`.

**Suggested change:**
```suggestion
                frames_to_remove = [k for k in output_dict[storage_key] if isinstance(k, int) and k < min_frame]
```

File: ultralytics/models/sam/predict.py [hunk 2]
```diff
@@ -3920,3 +3928,39 @@ def _drop_new_det_with_obj_limit(self, new_det_fa_inds, det_scores_np, num_to_ke
...
+                for f in frames_to_remove:
+                    del output_dict[storage_key][f]
+
+            # Also prune per-object non-cond outputs only
+            for obj_output in state.get("output_dict_per_obj", {}).values():
+                if storage_key in obj_output:
+                    frames_to_remove = [k for k in obj_output[storage_key] if k < min_frame]
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/predict.py]: ⚠️ **HIGH**: Same potential `TypeError` here if per-object `non_cond_frame_outputs` has any non-int keys. Filtering to int keys makes pruning robust.

**Suggested change:**
```suggestion
                    frames_to_remove = [k for k in obj_output[storage_key] if isinstance(k, int) and k < min_frame]
```

File: ultralytics/models/sam/predict.py [hunk 3]
```diff
@@ -3920,3 +3928,39 @@ def _drop_new_det_with_obj_limit(self, new_det_fa_inds, det_scores_np, num_to_ke
...
+                    for f in frames_to_remove:
+                        del obj_output[storage_key][f]
+
+    def _consolidate_tracker_states(self, tracker_states_local, max_states=4):
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/predict.py]: 💡 **MEDIUM**: `max_states` is unused, and the docstring says this prevents "unbounded growth" of `tracker_states_local`, but the implementation only removes empty states. Either enforce a cap (use `max_states`) or rename/remove the argument and adjust the docstring to match behavior.

File: ultralytics/models/sam/predict.py [hunk 4]
```diff
@@ -3920,3 +3928,39 @@ def _drop_new_det_with_obj_limit(self, new_det_fa_inds, det_scores_np, num_to_ke
...
+        """Prevent unbounded growth of tracker_states_local list."""
+        # Remove empty states
+        tracker_states_local[:] = [s for s in tracker_states_local if len(s.get("obj_ids", [])) > 0]
+        return tracker_states_local
+
+    def _prune_frame_scores(self, metadata, frame_idx, window=32):
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/predict.py]: 💡 **MEDIUM**: `window=32` duplicates `self.memory_bank_max_frames=32`. If either changes, pruning behavior becomes inconsistent. Consider tying this window to `self.memory_bank_max_frames` (or using a single config value) to keep all bounded-history bookkeeping aligned.

**Comment by Y-T-G** [comment]
@Laughing-q Do we actually need memory bank for semantic predictor?

**Comment by Laughing-q** [comment]
@Y-T-G yes we need that for the tracker part.
I think the key point here is that the SAM2VideoPredictor(we are using its extended `SAM3VideoPredictor` for the tracking part in `SAM3VideoSemanticPredictor`) is not pruning the `non_cond_frame_outputs`, probably we should add these pruning function in `SAM2VideoPredictor` and use it somewhere there. I believe we also have the same memory leak issue for SAM2 video tracking, it's just usually users only provide several prompts for objects while tracking so the memory leak issue did not get revealed before. But now with Semantic SAM3 model it's detecting much more objects while tracking videos so easily ran into the OOO issue.

EDIT: the `tracker_states_local` we are pruning in the function is actually a list that maintains multiple inference_state from `SAM3VideoPredictor`.

**Comment by Y-T-G** [comment]
There's still an issue where the memory increases as new objects appear and the inference keeps getting slower with each new object added, and doesn't decrease even when objects disappear.

**Comment by Laughing-q** [comment]
@Y-T-G yeah when there's new objects detected the tracker would add a new memory bank to maintain these new objects.

**Comment by Laughing-q** [comment]
@Y-T-G I think the original code already have a mechanism to clean up these objects that disappears here:
https://github.com/ultralytics/ultralytics/blob/2ddb06f7d22926c71a127ec0fc2ac7c3484d6945/ultralytics/models/sam/predict.py#L3330-L3330
so I removed the `_consolidate_tracker_states` function you introduced. Also I moved the `_prune_non_cond_memory` process in the `run_tracker_update_execution_phase` function where I think would be a better place to handle all the track status updating logic.
https://github.com/ultralytics/ultralytics/blob/2ddb06f7d22926c71a127ec0fc2ac7c3484d6945/ultralytics/models/sam/predict.py#L3331-L3333

EDIT: actually I found a better place here to put this pruning function in SAM3VideoPredictor:
https://github.com/ultralytics/ultralytics/blob/2ea866517e168145696a272d343838d9d3eefbf3/ultralytics/models/sam/predict.py#L2447-L2447
the same location you put in SAM2VideoPredictor.

**Review by Laughing-q — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
Merged! Huge thanks to @Y-T-G for driving this improvement and to @Laughing-q for the valuable contributions and collaboration on [PR #22969 (automatic SAM2 non-conditioning memory pruning)](https://api.github.com/repos/ultralytics/ultralytics/pulls/22969).

“As we look ahead into the next century, leaders will be those who empower others.” — Winston Churchill  
This PR embodies that idea by empowering users to run longer, steadier SAM2 video inference: bounding memory growth by pruning stale non-conditioning frame outputs, preserving the recent context that matters, and reducing RAM/VRAM creep that can lead to slowdowns or crashes.

Appreciate the thoughtful, practical defaults and the clean integration across `inference(...)` and `propagate_in_video(...)`—a real stability win for long-form tracking/segmentation workloads.

**Comment by Laughing-q** [comment]
@Y-T-G merged! nice work! I'm going to open another PR to attempt to fix the speed issue
